### PR TITLE
[CodeGen] Lower linalg.BatchMmt4DOp for non unit batch dimension.

### PR DESCRIPTION
**Pull request by Shukla-Gaurav**
This commit adds lowering of `linalg.BatchMmt4DOp` in case of non unit batch dimension using SCF Tiling utility. 
This change is a part of `DecomposeBatchMmt4DOpsPass` pass.

### CODE REVIEW

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
Isn't a simpler solution just modifying kernel dispatch to set tile size to 1 for batch dimension of batch_mmt4d?

**Comment by hanhanW** [comment]
> Isn't a simpler solution just modifying kernel dispatch to set tile size to 1 for batch dimension of batch_mmt4d?

We can do that, but it introduces one more entry in KernelConfig. This is the feature we still want because it gives us flexibility about distribution. I've been working on distribution heuristic and notice that a workgroup could have too less workload to process, especially for batch_matvec cases. It introduces "tiny dispatch". If we are able to add the support, we can have an unified distribution heuristic. It is something we want to have.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 1]
```diff
@@ -60,7 +61,23 @@ struct ConvertBatchMmt4DtoMmt4DPattern
...
-      return rewriter.notifyMatchFailure(op, "batch dim needs to be 1");
+      SmallVector<int64_t> tileSizes{1};
+      auto tilingInterfaceOp = cast<TilingInterface>(op.getOperation());
+      scf::SCFTileAndFuseOptions options{};
+      options.setTilingOptions(scf::SCFTilingOptions().setTileSizes(
+          getAsIndexOpFoldResult(op.getContext(), tileSizes)));
+      FailureOr<scf::SCFTileAndFuseResult> tiledResults =
+          scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
+              rewriter, tilingInterfaceOp, options);
+      if (failed(tiledResults))
```
Comment (MaheshRavishankar) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: Nit: Add `{` `}` around even single line statements (IREE diverges from LLVM style in this one way).

**Review by MaheshRavishankar — APPROVED** [review]
Thanks for the clarification.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 2]
```diff
@@ -60,7 +61,23 @@ struct ConvertBatchMmt4DtoMmt4DPattern
...
+        return rewriter.notifyMatchFailure(op, "unable to tile the operation");
+      SmallVector<Value> replacements;
+      replacements.resize(op.getNumResults());
+      for (const auto &[index, result] : llvm::enumerate(op.getResults())) {
+        replacements[index] = tiledResults->replacements.lookup(result);
+      }
+      rewriter.replaceOp(op, replacements);
+      return success();
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: I think it is better to decouple the logic from this pattern. We can either have a pattern to do the tiling or walkthrough batch_mmt4d ops (with funcOp.walk) and tile batch_dim with size=1. I prefer the latter one because we explicitly want to apply the pattern once. Can you move the logic to l.131 and do a func.walk? E.g.,

```
funcOp.walk([&](linalg::BatchMmt4DOp op) { ... } );
```

If we go with patterns, please create another pattern for this.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp [hunk 3]
```diff
@@ -405,6 +405,8 @@ void addVMVXDefaultPassPipeline(OpPassManager &passManager,
   addTileAndDistributePasses(passManager);
 
   if (enableMicrokernels) {
+    passManager.nest<ModuleOp>().nest<func::FuncOp>().addPass(
+        createDecomposeBatchMmt4DOpsPass());
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: I'd prefer adding this with an e2e test. It can be a separate PR. For this PR, let's focus on `Add support for tiling batch dimension with size=1 in batch_mmt4d decomposition`.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp (hunk 2 above)
Comment (Shukla-Gaurav) [reply compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: Moved it to func.walk. Thanks :)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp (hunk 3 above)
Comment (Shukla-Gaurav) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp]: Sure, will raise a separate patch for the this.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 4]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
...
   auto funcOp = getOperation();
 
+  funcOp.walk([&](linalg::BatchMmt4DOp op) {
+    auto out = op.getDpsInitOperand(0)->get();
+    auto outType = out.getType().cast<RankedTensorType>();
+    // Tile only non unit batch dimensions with tile size equals to 1.
+    if (outType.getShape()[0] <= 1) {
+      return;
+    }
+    SmallVector<int64_t> tileSizes{1};
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: Please use assignment syntax. The brace initialization with constructor syntax is dangerous. It could potentially lead to a bug, e.g.,

```
// Could invoke an intializer list constructor, or a two-argument constructor.
Frobber frobber{size, &bazzer_to_duplicate};

// Makes a vector of two doubles.
std::vector<double> fifty_pies{50, 3.14};
```

For more context, see https://abseil.io/tips/88

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 5]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
...
+    auto tilingInterfaceOp = cast<TilingInterface>(op.getOperation());
+    scf::SCFTileAndFuseOptions options{};
+    options.setTilingOptions(scf::SCFTilingOptions().setTileSizes(
+        getAsIndexOpFoldResult(ctx, tileSizes)));
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: [optional] This can be rewritten to below. The `setXXX` methods returns self at the end. It is a trick I learned from upstream authors. I think both are fine, so I mark it optional.

```cpp
auto options = scf::SCFTileAndFuseOptions().setTilingOptions( scf::SCFTilingOptions().setTileSizes(...) );
```

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 6]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
   MLIRContext *ctx = &getContext();
   auto funcOp = getOperation();
 
+  funcOp.walk([&](linalg::BatchMmt4DOp op) {
+    auto out = op.getDpsInitOperand(0)->get();
+    auto outType = out.getType().cast<RankedTensorType>();
+    // Tile only non unit batch dimensions with tile size equals to 1.
+    if (outType.getShape()[0] <= 1) {
+      return;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: We should return `WalkResult::advance();`, so it can keep the iteration. For failed cases, we return `WalkResult::interrupt();`; do the result check after function walk (e.g., `if (result.wasInterrupted()) ...`).

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 7]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
...
+    scf::SCFTileAndFuseOptions options{};
+    options.setTilingOptions(scf::SCFTilingOptions().setTileSizes(
+        getAsIndexOpFoldResult(ctx, tileSizes)));
+
+    IRRewriter rewriter(ctx);
+    FailureOr<scf::SCFTileAndFuseResult> tiledResults =
+        scf::tileConsumerAndFuseProducerGreedilyUsingSCFForOp(
+            rewriter, tilingInterfaceOp, options);
+    if (failed(tiledResults)) {
+      return;
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: I think we should signal the failure for this case. I.e., we should return `WalkResult::interrupt()`; check the result after function walk.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 8]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
...
+    }
+
+    SmallVector<Value> replacements;
+    replacements.resize(op.getNumResults());
+    for (const auto &[index, result] : llvm::enumerate(op.getResults())) {
+      replacements[index] = tiledResults->replacements.lookup(result);
+    }
+    rewriter.replaceOp(op, replacements);
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: We can simplify it to `rewriter.replaceOp(op, tiledResults->replacements[op.getResult()]);`.

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp [hunk 9]
```diff
@@ -130,6 +132,35 @@ void DecomposeBatchMmt4DOpsPass::runOnOperation() {
...
+      return;
+    }
+    SmallVector<int64_t> tileSizes{1};
+    auto tilingInterfaceOp = cast<TilingInterface>(op.getOperation());
+    scf::SCFTileAndFuseOptions options{};
+    options.setTilingOptions(scf::SCFTilingOptions().setTileSizes(
+        getAsIndexOpFoldResult(ctx, tileSizes)));
+
+    IRRewriter rewriter(ctx);
+    FailureOr<scf::SCFTileAndFuseResult> tiledResults =
```
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: nit: maybe rename it to `tileAndFuseResult`? There is a method (and a struct) for tiling only. I think renaming can make others have less confusion.

**Review by hanhanW — CHANGES_REQUESTED** [review]
Thanks, just few comments!

File: compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp (hunk 4 above)
Comment (Shukla-Gaurav) [reply compiler/src/iree/compiler/Codegen/Common/DecomposeBatchMmt4DOps.cpp]: Thanks:)

**Review by hanhanW — APPROVED** [review]
